### PR TITLE
Pluralization support

### PR DESCRIPTION
**Pull request by JoelBesada**
Add support for strings that use the pluralization feature of i18n-js. From this input: 
```
{
  "common": {
    "items": {
      "one": "One item.",
      "other": "{{count}} items."
  }
}
```
We don't want to interpret this as two strings `"common.items.one"` and `"common.items.other"` as it does currently, but rather collapse them into one string `"common.items"` with a `count` parameter, returning either `"One item."` or ` "{{count}} items."`. 

To do this, we assume that a key that contains an object with *only* keys `"zero"`, `"one"` or  `"other"` to be in pluralized form. This might however end up being a breaking change if someone was previously using this library and not expecting these keys to be collapsed into the parent as they will with this PR.
 
We needed to add this feature to make this library fit our needs, and it's been working out great. Happy to hear if you'll consider merging this so we can drop our fork. :)

### CODE REVIEW

**Comment by hotchemi** [comment]
let me take a look, thank you!

File: src/lib/__tests__/astTest.ts [hunk 1]
```diff
@@ -39,5 +39,19 @@ describe('ast', () => {
       );
       expect(actual).toEqual(expected);
     });
+
+    it('returns dts file with union types', () => {
```
Comment (hotchemi) [inline src/lib/__tests__/astTest.ts:43]: Could you add one more test case to prove that it works when `value` does not contain the phrase which requires {{count}} ? 🙇

File: src/lib/__tests__/expected/union-types.d.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    function currentLocale(): string;
+    function t(key: "common.items", opts: {
+        count: any;
+    }): "One item." | "{{count}} items.";
+}
+
+declare module "*.json" {
+    const value: any;
+    export default value;
+}
```
Comment (hotchemi) [inline src/lib/__tests__/expected/union-types.d.ts]: Nit
```suggestion
}

```

File: src/lib/__tests__/parserTest.ts [hunk 3]
```diff
@@ -16,6 +16,13 @@ describe('parser', () => {
         'test1',
       ]);
     });
+
+    it('supports both interpolation syntaxes', () => {
```
Comment (hotchemi) [inline src/lib/__tests__/parserTest.ts:20]: Thx, btw is this related to pluralization?

File: src/lib/__tests__/parserTest.ts [hunk 4]
```diff
@@ -46,9 +53,55 @@ describe('parser', () => {
...
+          value: ['One {{type}} item.', 'Many {{type}} items.'],
+        },
+      ]);
+    });
+  });
+
+  describe('isPluralized', () => {
+    it('returns true for objects that only contain pluralization keys', () => {
+      expect(
+        isPluralized({
```
Comment (hotchemi) [inline src/lib/__tests__/parserTest.ts:106]: To cover fully, perhaps we need to make sure the following?

- one -> true
- other -> true
- zero -> true
- one, other -> true
- one, zero -> true
- other, zero -> true

**Review by hotchemi — COMMENTED** [review]
Looks great 👍 About backward compatibility I deem it's bearable as i18-js supports pluralization and we'll just follow that(might need to bump up a major version though).

Btw just found i18n-js supports some exceptional lang rule(like Russian). presumably it'd be great if we can support it as well(haven't thought so much) 😉 
Screen Shot 2020-06-14 at 18 42 16

https://github.com/fnando/i18n-js

**Comment by hotchemi** [comment]
@yuya-takeyama @christineoo sorry for mentioning, could you take a look when you have time(and merge)? 🙇 I don't have ownership anymore.

File: src/lib/__tests__/parserTest.ts (hunk 3 above)
Comment (JoelBesada) [reply src/lib/__tests__/parserTest.ts:20]: Not really, no. Just noticed this while adding the pluralization support, sorry for bundling this in with the same PR. 😅

Comment (hotchemi) [reply src/lib/__tests__/parserTest.ts:20]: noted, thanks a lot😃

**Comment by JoelBesada** [comment]
> Btw just found i18n-js supports some exceptional lang rule(like Russian). presumably it'd be great if we can support it as well(haven't thought so much) 😉

I'll see if I can add a CLI flag to override the `PLURALIZATION_KEYS` value, that way you should be able to customize according to any extra rules you've added yourself.

**Comment by hotchemi** [comment]
@JoelBesada thank you so much!
Btw, the current situation is the library has not been used so much internally(and I've already left Quipper and it nearly means there's no active maintainer in the company). Considering that situation if you're using the library actively you can just fork and continue to maintain/develop😉 Then we'll just archive the repo and update README to deprecate it.

**Comment by JoelBesada** [comment]
> @JoelBesada thank you so much!
> Btw, the current situation is the library has not been used so much internally(and I've already left Quipper and it nearly means there's no active maintainer in the company). Considering that situation if you're using the library actively you can just fork and continue to maintain/develop😉 Then we'll just archive the repo and update README to deprecate it.

I've updated the PR adding in a CLI option to override the pluralization keys. 

Regarding taking over the active development of this library through our fork, I'm not sure we're going to continue adding much more after these changes, which would mean our fork wouldn't be that actively maintained either. I appreciate the offer, and we'll reconsider it again if we find ourselves making more additions to the library, but for now I think it would be best to just have this  update merged and pushed as a new version to NPM within this repo if possible.

**Comment by hotchemi** [comment]
Thank you, let us take a look.

**Review by hotchemi — APPROVED** [review]
Looks good to me👍
Thank you so much!

**Review by yuya-takeyama — APPROVED** [review]
Sorry for the delay, LGTM 👍

**Comment by yuya-takeyama** [comment]
@JoelBesada 
Your change has been released as 0.3.0.
https://www.npmjs.com/package/i18n-dts/v/0.3.0

Thanks for your contribution!
